### PR TITLE
V3

### PR DESCRIPTION
**Pull request by dimitrisnl**
Change-type: major

### CODE REVIEW

**Comment by dimitrisnl** [comment]
Example with the most minimal setup.

https://dimitrisnl.github.io/landr-test/
https://github.com/dimitrisnl/landr-test

**Comment by dimitrisnl** [comment]
My latest changes introduced this error while running npx

```
dr: line 1: syntax error near unexpected token `('
... line 1: `const capitano = require('capitano');'
```

Investigating

**Comment by dimitrisnl** [comment]
@jviotti 	@LucianBuzzo 	I introduced capitano on the project, which allows having both the `deploy` option, as well as the `preview` mode.

I believe it's good to have since we will inevitably want to pass more params. For example having a different port for the local server.

**Comment by jviotti** [comment]
@dimitrisnl Sounds good to me. The PR LGTM for me if the commands are:

- `landr build`: Runs scrutinizer, and generates a website out of it
- `landr deploy`: Deploys whatever `landr build` generated, calling `landr build` if it never ran before

Then I guess `landr preview` opens your browser at the entry point that `landr build` generates?

**Comment by dimitrisnl** [comment]
exactly @jviotti 	. It stats a local server to preview the changes. I will introduce the `landr build` command, since now it's part of both commands.

**Comment by dimitrisnl** [comment]
Why would we want to separate the build command from the preview/deploy ones?
In case people don't want to publish in GH-pages?

**Comment by jviotti** [comment]
@dimitrisnl Ah, I was thinking that `build` would be the way to test it out before publishing, but if `preview` does that already then it's fine.

**Comment by dimitrisnl** [comment]
Nvm @jviotti 	I implemented it, and cleaned up the code. It's good to have it.

**Comment by LucianBuzzo** [comment]
Installing the latest prerelease build failed for me with the following error:
```
ಠ_ಠ (08:46 am) hydroid/ (master) npm install landr@3.0.0-v3-67efb4b6938bd7fe52c087463f17599de3e7230a
npm ERR! code E404
npm ERR! 404 Not Found: har-validator@https://registry.npmjs.org/har-validator/-/har-validator-5.1.2.tgz

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/lucianbuzzo/.npm/_logs/2019-06-17T07_47_22_078Z-debug.log
```

Node version: v9.10.0
npm version: 5.6.0

**Comment by LucianBuzzo** [comment]
The `preview` command just shows me a screen like this:
Screen Shot 2019-06-17 at 09 01 57

**Comment by LucianBuzzo** [comment]
When running the `deploy` command, it appeared to fail with the following error, then it asked me to login to GitHub:
```
Fetching data: 0% ...
Fetching data: 11% ...
Fetching data: 22% ...
Fetching data: 33% ...
Fetching data: 44% ...
Fetching data: 55% ...
Fetching data: 66% ...
Fetching data: 77% ...
Fetching data: 88% ...
Removing node_modules
Installing dependencies
yarn install v1.15.2
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
warning "rendition > mermaid > dagre-d3-renderer@0.5.8" has incorrect peer dependency "d3@^4.13.0".
[4/4] Building fresh packages...
Done in 17.50s.
Building
yarn run v1.15.2
$ react-static build

=> Info: No 'siteRoot' is defined in 'static.config.js'. This is suggested for absolute url's and a sitemap.xml to be automatically generated.

=> Building Routes...
=> [✓] Routes Built: 4.296ms
=> Copying public directory...
=> [✓] Public directory copied: 0.667ms
=> Bundling App...
=> [✓] App Bundled: 23995.118ms
=> Fetching Site Data...
=> [✓] Site Data Downloaded: 0.064ms
=> Fetching Route Data...

=> [✓] Route Data Downloaded: 1.461ms
=> Exporting Route Data...
=> [✓] Route Data Exported: 0.265ms
=> Exporting HTML...

  TypeError: Failed exporting HTML for URL / (src/containers/Home): Cannot read   property 'asssets' of null
  
  - static.27fe3487.js:291 
    [icecave]/[landr]/theme/dist/static.27fe3487.js:291:92
  
  - react-dom-server.node.development.js:2792 processChild
    [icecave]/[theme]/[react-dom]/cjs/react-dom-server.node.development.js:2792:    14
  
  - react-dom-server.node.development.js:2716 resolve
    [icecave]/[theme]/[react-dom]/cjs/react-dom-server.node.development.js:2716:    5
  
  - react-dom-server.node.development.js:3100 ReactDOMServerRenderer.render
    [icecave]/[theme]/[react-dom]/cjs/react-dom-server.node.development.js:3100:    22
  
  - react-dom-server.node.development.js:3059 ReactDOMServerRenderer.read
    [icecave]/[theme]/[react-dom]/cjs/react-dom-server.node.development.js:3059:    29
  
  - react-dom-server.node.development.js:3526 renderToString
    [icecave]/[theme]/[react-dom]/cjs/react-dom-server.node.development.js:3526:    27
  
  - index.js:519 renderToStringAndExtract
    [icecave]/[theme]/[react-static]/lib/static/index.js:519:68
  
  - static.config.js:27 Object.renderToHtml
    [icecave]/[landr]/theme/static.config.js:27:18
  
  - index.js:556 _callee8$
    [icecave]/[theme]/[react-static]/lib/static/index.js:556:39
  
  - runtime.js:62 tryCatch
    [icecave]/[theme]/[regenerator-runtime]/runtime.js:62:40
  

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: README.md [hunk 1]
```diff
@@ -1,66 +1,44 @@
...
-## Why landr
-
-You have to maintain your source code why maintain a website too?
-
-As a software company we have a growing number of websites to build and maintain. We built landr so we could focus on our projects and not their websites.
+3. Run the CLI
+`GITHUB_TOKEN=[your token] npm run landr`
 
-Most OS websites the same, the have a hero, a getting started and some docs. There is definitely room for automation.
+1. Landr will push the changes to the `gh-pages` branch.
```
Comment (LucianBuzzo) [inline README.md]: The phrase `the changes` is quite ambiguous, you could probable change this to 
```suggestion
1. Landr will push the generated files to the `gh-pages` branch.
```

File: README.md [hunk 2]
```diff
@@ -1,66 +1,44 @@
...
+<h1 align="center">
+    Landr
+</h1>
 
-source code = website!
+## 🚀 Quick start
 
-> Build a website for your software projects with one command.
+1. Navigate to the repository
+2. run `GITHUB_TOKEN=[your token] npx landr deploy`
```
Comment (LucianBuzzo) [inline README.md]: Can you clarify what permissions the token needs?

File: README.md [hunk 3]
```diff
@@ -1,66 +1,44 @@
...
-Build site:
-```
-landr build
-```
-
-View built site locally:
-```
-landr serve
-```
+1. Add the a npm script entry in your `package.json` file
```
Comment (LucianBuzzo) [inline README.md]:
```suggestion
1. Add the following to the script entry in the project's `package.json` file
```

File: README.md [hunk 4]
```diff
@@ -1,66 +1,44 @@
-# landr
+<h1 align="center">
+    Landr
+</h1>
```
Comment (LucianBuzzo) [inline README.md]: A short explanation fo what landr does would be ideal here

File: bin/index.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+capitano.command({
+  signature: 'deploy',
+  description: 'Build & publish the website in Github Pages',
+  action: async ({ help }) => {
+    if (help) {
+      showHelp();
+      process.exit(1);
+    }
+
+    await build();
```
Comment (LucianBuzzo) [inline bin/index.js]: If the build fails does this process exit?

**Comment by dimitrisnl** [comment]
@LucianBuzzo 	Can you share with me which repository you tried? 

This should be a scrutinizer bug.

File: bin/utils.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+const chalk = require('chalk');
+const gitInfo = require('gitinfo');
+
+const { log } = console;
+const rootPath = process.cwd();
+const landrPath = path.resolve(__dirname, '..');
+
+const runFromLandr = () => shell.cd(landrPath);
+const runFromTheme = () => shell.cd(path.resolve(landrPath, 'theme'));
+const isUsingYarn = () => !!shell.which('yarn');
```
[the remark this answers is not in the document]
Comment (dimitrisnl) [reply bin/utils.js]: I think of dropping this check altogether and to just use NPM.

File: bin/index.js (hunk 5 above)
Comment (dimitrisnl) [reply bin/index.js]: Yes, it does

File: README.md (hunk 2 above)
Comment (dimitrisnl) [reply README.md]: Great suggestion. Also opened this one https://github.com/balena-io-modules/scrutinizer/issues/56 👍

File: README.md (hunk 4 above)
Comment (dimitrisnl) [reply README.md]: The Readme needs some love for sure. For now, I added the previous subtitle "Build a website for your software projects with one command.".

File: bin/utils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (LucianBuzzo) [reply bin/utils.js]: What exactly do you use npm/yarn for here?

**Comment by LucianBuzzo** [comment]
The `landr deploy` command asks me for a GitHub username even though I provide a token

File: bin/utils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (dimitrisnl) [reply bin/utils.js]: Installing the react-static dependencies (Rendition, React, etc).

**Comment by dimitrisnl** [comment]
@LucianBuzzo 	Can you give more context on your use case? I successfully built a website out of any repo I tried (apart from the ones without tags, thanks for spotting that).

Could this be a private repo?

**Comment by LucianBuzzo** [comment]
@dimitrisnl I initially tried on this repo https://github.com/LucianBuzzo/icecave - I just tried again to see if its still an issue and the deploy went fine, so I think you can ignore my comment

File: bin/utils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (LucianBuzzo) [reply bin/utils.js]: @dimitrisnl I can live with this as a temporary abstraction, but this is a really ugly approach and we need to find a more elegant way of building the site.

**Comment by LucianBuzzo** [comment]
After running the deploy command, I see :
```
Done in 33.08s.
Website successfully build and deployed 🏠`
```
But when I look at the repo https://github.com/LucianBuzzo/icecave/branches I don't see a `gh-pages` branch or any new commits.
Here is my git config:
```
[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
	ignorecase = true
	precomposeunicode = true
[remote "origin"]
	url = git@github.com:LucianBuzzo/icecave.git
	fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
	remote = origin
	merge = refs/heads/maste
```

File: bin/deploy.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  runFromTheme(); // move to the directory of React Static
+  const githubUrl = getGithubUrl();
+
+  await ghPages.publish('dist', {
+    message: 'Deployed by landr 🏠',
+    branch: 'gh-pages',
+    repo: githubUrl,
+  });
+
+  print.success('Website successfully build and deployed 🏠`');
```
Comment (LucianBuzzo) [inline bin/deploy.js]: NIT: there are some typos here
```suggestion
  print.success('Website successfully build and deployed 🏠');
```

**Comment by dimitrisnl** [comment]
@LucianBuzzo 	

https://github.com/dimitrisnl/icecave/tree/gh-pages
https://dimitrisnl.github.io/icecave/

This is a bit weird. Btw is your last line a typo? `merge = refs/heads/maste`

**Comment by dimitrisnl** [comment]
Btw the lack of releases are due to scrutinizer

```
  "latestRelease": null,
  "latestPreRelease": null
```

@nazrhom 	Any clue on why this repo with tags & releases have no latest Release populated? https://github.com/dimitrisnl/icecave

**Comment by LucianBuzzo** [comment]
@dimitrisnl Yeah thats a typo, my git config has the correct last line `merge = refs/heads/master`

**Comment by dimitrisnl** [comment]
I suspect because https://lucianbuzzo.github.io/icecave/ is already active, that GH pages are working and build out of master

Here's my version (with a release from GH)

**Comment by dimitrisnl** [comment]
I'm having a very hard time figuring this bug out. The deploy script is ported from the v2 codebase.

**Comment by dimitrisnl** [comment]
- https://dimitrisnl.github.io/postcss-autocorrect/
- https://dimitrisnl.github.io/landr-test/
- https://dimitrisnl.github.io/icecave/
- https://dimitrisnl.github.io/Happy-React-switch/

**Comment by jviotti** [comment]
@dimitrisnl Can we merge this PR so I can take it as a starting point? Navigating the code through a PR is much harder :P

**Pull request merged**